pep-0xxx.rst

```diff
@@ -0,0 +1,1025 @@
+
+    "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256"
+
+That makes parsing a complete OpenSSL cipher string extremely tricky. Add to
+the fact that there are other meta-characters, such as "!" (exclude all cipher
+suites that match this criterion, even if they would otherwise be included:
+"!MD5" means that no cipher suites using the MD5 hash algorithm should be
+included), "-" (exclude matching ciphers if they were already included, but
+allow them to be re-added later if they get included again), and "+" (include
+the matching ciphers, but place them at the end of the list), and you get an
+*extremely* complex format to parse. On top of this complexity it should be
+noted that the actual result depends on the OpenSSL version, as an OpenSSL
+cipher string is valid so long as it contains at least one cipher that OpenSSL
+recognises.
+
+OpenSSL also uses different names for its ciphers than the names used in the
+relevant specifications. See the manual page for ``ciphers(1)`` for more
+details.
+
+The actual API inside OpenSSL for the cipher string is simple:
+
+    char *cipher_list = <some cipher list>;
+    int rc = SSL_CTX_set_cipher_list(context, cipher_list);
+
+This means that any format that is used by this module must be able to be
+converted to an OpenSSL cipher string for use with OpenSSL.
+
+SecureTransport
+^^^^^^^^^^^^^^^
+
+SecureTransport is the macOS system TLS library. This library is substantially
+more restricted than OpenSSL in many ways, as it has a much more restricted
+class of users. One of these substantial restrictions is in controlling
+supported cipher suites.
+
+Ciphers in SecureTransport are represented by a C ``enum``. This enum has one
+entry per cipher suite, with no aggregate entries, meaning that it is not
+possible to reproduce the meaning of an OpenSSL cipher string like
+"ECDH+AESGCM" without hand-coding which categories each enum member falls into.
+
+However, the names of most of the enum members are in line with the formal
+names of the cipher suites: that is, the cipher suite that OpenSSL calls
+"ECDHE-ECDSA-AES256-GCM-SHA384" is called
+"TLS_ECDHE_ECDHSA_WITH_AES_256_GCM_SHA384" in SecureTransport.
+
+The API for configuring cipher suites inside SecureTransport is simple:
+
+    SSLCipherSuite ciphers[] = {TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, ...};
+    OSStatus status = SSLSetEnabledCiphers(context, ciphers, sizeof(cphers));
+
+SChannel
+^^^^^^^^
+
+SChannel is the Windows system TLS library.
+
+SChannel specifies cipher suites not by using symbolic constants but by using
+textual names for the cipher suites (the type used in the API is ``LPWSTR``,
+which is a Windows typedef for a pointer to a null-terminated string of
+16-bit unicode characters). The list of supported cipher suites is provided in
+`MSDN articles`_, and can be dynamically found at runtime by using the
+``BCryptEnumContextFunctions`` function appropriately.
```

Comment:
You can only specify specific cipher suites to enable and disable for the whole machine. You'll notice at https://msdn.microsoft.com/en-us/library/windows/desktop/bb870930(v=vs.85).aspx#adding__removing__and_prioritizing_cipher_suites, the function it called for the `CRYPT_LOCAL` config table, affecting the whole machine. You'll see some similar discussion at http://stackoverflow.com/questions/30193707/how-can-i-add-remove-cipher-suites-from-schannel-without-affecting-other-proces.

You have some control over allowable algorithms via the `palgSupportedAlgs` field of the `SCHANNEL_CRED ` struct that is passed to `AcquireCredentialsHandle()`. Additionally, the constant `SCH_USE_STRONG_CRYPTO` can help disable "weak" crypto, as defined by Microsoft. This allows you to blacklist certain things, but you aren't ever going to have control over specific cipher suites. You can see an implementation using this at https://github.com/wbond/oscrypto/blob/0.17.3/oscrypto/_win/tls.py#L193-L272.

In general, SecureTransport and SChannel have far more restricted supported cipher suites than OpenSSL does. I didn't bother trying to whitelist cipher suites with SecureTransport, but instead created a blacklist and removed anything with a poor algorithm from `SSLGetSupportedCiphers ()`. https://github.com/wbond/oscrypto/blob/0.17.3/oscrypto/_osx/tls.py#L617-L657.